File: chapter-15/recipe-03/cxx-example/src/buggy.cpp
#include "buggy.hpp"

#include <iostream>

int function_leaky() {

  double *my_array = new double[1000];

  // do some work ...

  // we forget to deallocate the array
  // delete[] my_array;

  return 0;
}

int function_use_after_free() {

  double *another_array = new double[1000];

  // do some work ...

  // deallocate it, good!
  delete[] another_array;

  // however, we accidentally use the array
  // after it has been deallocated
  std::cout << "not sure what we get: " << another_array[123] << std::endl;

  return 0;
}

File: chapter-14/recipe-03/fortran-example/src/buggy.f90
module buggy

  implicit none

  public function_leaky

  private

contains

  integer function function_leaky()
    real(8), pointer :: my_array(:)
    allocate(my_array(1000))
    ! accidentally allocate twice
    allocate(my_array(1000))

    ! do some work ...

    deallocate(my_array)

    function_leaky = 0
  end function

end module
